### PR TITLE
Refactor PackageGraphWatcher

### PR DESCRIPTION
**Pull request by natebosch**
Use `StreamGroup.merge` to combine all the individual node watchers
instead of dropping down to StreamController directly. This saves us
from needing to handle things like cancel forwarding.

- Make the `node` field public on `PackageNodeWatcher` so we can more
  easily `.map()` over the instances and set up the filtering.
- Add a `_nestedPathFilter` helper method to get a closure over the
  nested paths without needing a separate variable higher up.
- Fix the test for nested path filtering. The old test did not exercise
  the filtering code path since it did not send the change through the
  "outer" watcher like what would happen in a real file system.
- Enforce that a single call is made to `watch()` explicitly.
- Remove the `relative` concept from the node watcher since it was
  unused outside of tests.

### CODE REVIEW

**Comment by natebosch** [comment]
There is a behavior change here in that the listeners are started on the call to `watch()` while before they were started when the stream was listened to. In practice this shouldn't make much difference since we don't call `watch()` unless we intend to listen to the stream.

**Comment by natebosch** [comment]
I take that back - it's really easy to retain the old behavior with a `LazyStream` so I may as well do that.

File: build_runner/lib/src/watcher/graph_watcher.dart [hunk 1]
```diff
@@ -38,57 +39,34 @@ class PackageGraphWatcher {
...
-      onListen: () {
-        subscriptions = logTimedSync(
-          _logger,
-          'Setting up file watchers',
-          () => _watch(controller),
-        );
-      },
-      onCancel: () {
-        for (final subscription in subscriptions) {
-          subscription.cancel();
```
Comment (grouma) [inline build_runner/lib/src/watcher/graph_watcher.dart]: Does lazy stream cancel subscriptions upon cancel?

File: build_runner/lib/src/watcher/graph_watcher.dart [hunk 2]
```diff
@@ -38,57 +39,34 @@ class PackageGraphWatcher {
...
-        // TODO: Consider a faster filtering strategy.
-        if (nestedPackages.any((path) => event.id.path.startsWith(path))) {
-          return;
-        }
-        sink.add(event);
-      }));
-    });
+  Stream<AssetChange> _watch() {
+    final allWatchers = _graph.allPackages.values
+        .where((node) => !(node.dependencyType == DependencyType.hosted ||
```
Comment (jakemac53) [inline build_runner/lib/src/watcher/graph_watcher.dart]: Not directly related, but we may want to consider changing this to just `node.dependencyType == DependencyType.path`? For instance the there are more dependency types, like SDK (for the flutter package).

File: build_runner/lib/src/watcher/graph_watcher.dart (hunk 1 above)
Comment (natebosch) [reply build_runner/lib/src/watcher/graph_watcher.dart]: Yes

File: build_runner/lib/src/watcher/graph_watcher.dart (hunk 2 above)
Comment (natebosch) [reply build_runner/lib/src/watcher/graph_watcher.dart]: Good idea

**Pull request merged**